ui/src/features/Editor/useCanvasCopyPaste.ts

```diff
@@ -221,46 +249,176 @@ export default ({
     [],
   );
 
+  const processNodesAndEdges = useCallback(
+    (selectedNodes: Node[], selectedEdges: Edge[]) => {
+      const processedNodeIds = new Set();
+      const nodesToProcess = [...selectedNodes];
+      const edgesToProcess = [...selectedEdges];
+
+      selectedNodes.forEach((node) => {
+        processedNodeIds.add(node.id);
+      });
+
+      const batchNodeIds = selectedNodes
+        .filter((node) => node.type === "batch")
+        .map((node) => node.id);
+
+      nodes.forEach((node) => {
+        if (
+          node.parentId &&
+          batchNodeIds.includes(node.parentId) &&
+          !processedNodeIds.has(node.id)
+        ) {
+          nodesToProcess.push(node);
+          processedNodeIds.add(node.id);
+        }
+      });
+
+      const allNodeIds = new Set(nodesToProcess.map((node) => node.id));
+
+      const processedEdgeIds = new Set(selectedEdges.map((edge) => edge.id));
+
+      edges.forEach((edge) => {
+        if (
+          (allNodeIds.has(edge.source) || allNodeIds.has(edge.target)) &&
+          !processedEdgeIds.has(edge.id)
+        ) {
+          edgesToProcess.push(edge);
+          processedEdgeIds.add(edge.id);
+        }
+      });
+
+      return {
+        nodes: nodesToProcess,
+        edges: edgesToProcess,
+      };
+    },
+    [nodes, edges],
+  );
```

Comment:
I think a lot more is going on in here than necessary. Maybe we will leave it as-is, but want to walk through it with you first later